### PR TITLE
[API][Product] Product sorting by name

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Screenshot 2020-03-31 at 14 40 59

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorLocalesContext.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+declare(strict_types=1);
+
+namespace Sylius\Behat\Context\Ui\Admin;
+
+use Behat\Behat\Context\Context;
+use Sylius\Behat\Page\Admin\Administrator\UpdatePageInterface;
+use Sylius\Behat\Service\SharedStorageInterface;
+use Sylius\Component\Core\Model\AdminUserInterface;
+
+final class ManagingAdministratorLocalesContext implements Context
```
Comment (Tomanhez) [inline src/Sylius/Behat/Context/Ui/Admin/ManagingAdministratorLocalesContext.php]: I added it because I had conflict with contexts in suite

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 2]
```diff
@@ -44,18 +48,40 @@ final class ManagingProductsContext implements Context
     /** @var SharedStorageInterface */
     private $sharedStorage;
 
+    /** @var array */
+    private $sortTypes;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php]: It should rather be a const

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/product.yml [hunk 3]
```diff
@@ -7,7 +7,7 @@ sylius_grid:
                     class: "%sylius.model.product.class%"
                     repository:
                         method: createListQueryBuilder
-                        arguments: ["%locale%", $taxonId]
+                        arguments: ["expr:service('security.token_storage').getToken().getUser().getLocaleCode()", $taxonId]
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/product.yml]: Shouldn't it be taken from locale context?

File: src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                $queryBuilder
+                    ->addSelect('translation')
+                    ->innerJoin(
+                        'o.translations',
+                        'translation',
+                        'WITH',
+                        'translation.locale = :locale'
+                    )
+                    ->orderBy('translation.name', $direction)
+                    ->setParameter('locale', $value['localeCode']);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php]:
```suggestion
                    ->setParameter('locale', $value['localeCode'])
                ;
```

File: src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+                return;
+            }
+
+            $queryBuilder
+                ->addSelect('translation')
+                ->innerJoin('o.translations', 'translation')
+                ->orderBy('translation.name', $direction);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php]:
```suggestion
                ->orderBy('translation.name', $direction)
            ;
```

File: src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+{
+    protected function filterProperty(
+        string $property,
+        $value,
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        string $operationName = null
+    ): void {
+        if ($property === 'order') {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php]: IMHO, it would be more readable to reverse this condition

File: src/Sylius/Behat/Client/ApiClientInterface.php [hunk 7]
```diff
@@ -34,7 +34,7 @@ public function delete(string $id): Response;
 
     public function filter(): Response;
 
-    public function sort(string $field, string $value): Response;
+    public function sort(array $sorting): Response;
```
Comment (GSadee) [inline src/Sylius/Behat/Client/ApiClientInterface.php]: Why did you change that?

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 8]
```diff
@@ -29,6 +30,9 @@
 
 final class ManagingProductsContext implements Context
 {
+    /** @var ApiClientInterface */
+    private $adminClient;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php]: Add this as the second argument, because `$client` is the default API client for this context and it should be imo the first one
```suggestion
    private $adminUsersClient;
```
or
```suggestion
    private $administratorsClient;
```

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 9]
```diff
@@ -44,18 +48,40 @@ final class ManagingProductsContext implements Context
...
+    public function iSwitchTheLocaleToTheLocale(string $localeCode): void
+    {
+        /** @var AdminUserInterface $adminUser */
+        $adminUser = $this->sharedStorage->get('administrator');
+
+        $this->adminClient->buildUpdateRequest((string) $adminUser->getId());
+
+        $this->adminClient->updateRequestData(['localeCode' => $localeCode]);
+
+        $this->adminClient->update();
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php]: Unnecessary blank lines imo

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 10]
```diff
@@ -526,19 +542,14 @@ public function productWithNameShouldNotBeAdded(string $field, string $value): v
         Assert::false($this->hasProductWithFieldValue($this->client->index(), $field, $value));
     }
 
-    private function hasProductImage(Response $response, ProductInterface $product): bool
+    private function getAdminLocaleCode(): string
     {
-        $productFromResponse = $this->responseChecker->getResponseContent($response);
+        /** @var AdminUserInterface $admin */
+        $admin = $this->sharedStorage->get('administrator');
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php]:
```suggestion
        $adminUser = $this->sharedStorage->get('administrator');
```
or
```suggestion
        $administrator = $this->sharedStorage->get('administrator');
```

File: src/Sylius/Behat/Client/ApiClientInterface.php (hunk 7 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Client/ApiClientInterface.php]: I changed it because i need add next filter for locale, i can add next variable for locale and default add null but in ApiClient I would have to check null and add few lines for version with and without locale.

Screenshot 2020-03-31 at 20 46 20

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 11]
```diff
@@ -29,6 +30,11 @@
 
 final class ManagingProductsContext implements Context
 {
+    public const SORT_TYPES = ['ascending' => 'asc', 'descending' => 'desc'];
+
+    /** @var ApiClientInterface */
+    private $adminUserClient;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php:36]:
```suggestion
    private $adminUsersClient;
```
as I wrote yesterday 😢

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 12]
```diff
@@ -45,19 +51,35 @@ final class ManagingProductsContext implements Context
     private $sharedStorage;
 
     public function __construct(
+        ApiClientInterface $adminUserClient,
         ApiClientInterface $client,
         ApiClientInterface $productReviewsClient,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php:56]: Order should be changed as the $client is the default one for this context, I wrote it in review yesterday 😢
And here you can see that there is already some convention for naming api client services, but you added it differently..

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 13]
```diff
@@ -548,20 +555,35 @@ private function getFieldValueOfFirstProduct(array $product, string $field): ?st
...
         return null;
     }
 
+    private function hasProductImage(Response $response, ProductInterface $product): bool
+    {
+        $productFromResponse = $this->responseChecker->getResponseContent($response);
+
+        if (
+            isset($productFromResponse['images']) &&
+            isset($productFromResponse['images'][0]) &&
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php:570]:
```suggestion
            isset($productFromResponse['images'][0]) &&
```
Should work exactly same

File: src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php [hunk 14]
```diff
@@ -548,20 +555,35 @@ private function getFieldValueOfFirstProduct(array $product, string $field): ?st
...
+            $productFromResponse['images'][0]['path'] === $product->getImages()->first()->getPath()
+        ) {
+            return true;
+        }
+
+        return false;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Admin/ManagingProductsContext.php:576]: Or even:
```suggestion
        return 
            isset($productFromResponse['images'][0]) &&
            $productFromResponse['images'][0]['path'] === $product->getImages()->first()->getPath()
        ;
```

File: src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                return;
+            }
+
+            $direction = $value['translation.name'];
+
+            if (isset($value['localeCode'])) {
+                $queryBuilder
+                    ->addSelect('translation')
+                    ->innerJoin(
+                        'o.translations',
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Filters/TranslationOrderNameFilter.php:42]: We cannot be sure, that `o` will be used here.
```suggestion
                        sprintf('%s.translations', $queryBuilder->getRootAliases()[0]));
```

File: src/Sylius/Bundle/AdminBundle/Resources/config/grids/product.yml [hunk 16]
```diff
@@ -7,7 +7,7 @@ sylius_grid:
                     class: "%sylius.model.product.class%"
                     repository:
                         method: createListQueryBuilder
-                        arguments: ["%locale%", $taxonId]
+                        arguments: ["expr:service('sylius.context.locale.admin_based').getLocaleCode()", $taxonId]
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/config/grids/product.yml:10]: What will happen, if we will call:
```suggestion
                        arguments: ["expr:service('sylius.context.locale').getLocaleCode()", $taxonId]
```

**Comment by GSadee** [comment]
Thank you, Tomasz! :1st_place_medal:
